### PR TITLE
Refine training data preparation and add Tulu V2 mixture.

### PR DESCRIPTION
**Pull request by yizhongw**

### CODE REVIEW

File: open_instruct/reformat_datasets.py [hunk 1]
```diff
@@ -515,37 +547,243 @@ def convert_open_orca_data(data_dir, output_dir, num_gpt4_examples=100000, num_g
...
+                "id": f"science_{idx}",
+                "messages": [
+                    {"role": "user", "content": example["input"]},
+                    {"role": "assistant", "content": example["output"]}
+                ],
+            }) + "\n")
 
+
+
+def should_be_filtered(example):
```
Comment (yizhongw) [inline open_instruct/reformat_datasets.py:592]: @hamishivi This function does the filtering words check.

File: open_instruct/reformat_datasets.py [hunk 2]
```diff
@@ -370,12 +388,13 @@ def convert_baize_data(data_dir, output_dir):
...
 
-def convert_oasst1_data(data_dir, output_dir):
+def convert_oasst1_data(data_dir, output_dir, top_k_reply=None):
     '''
     For OASST1, because it's in a tree structure, where every user input might get multiple replies, 
     we have to save every path from the root node to the assistant reply (including both leaf node and intemediate node).
     This results in some of the messages being duplicated among different paths (instances).
-    Be careful when using this dataset for training. Ideally, you should only minimize the loss of the last message in each path.
+    You can set top_k_reply to control how many replies to consider when traversing the tree, which will consider the replies with 
+    the highest human-rewviewed quality scores.
```
Comment (hamishivi) [inline open_instruct/reformat_datasets.py]:
```suggestion
    the highest human-reviewed quality scores.
```

File: scripts/prepare_science_data.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+names = [
+    "evidence_inference",
+    "qasper_truncated_4000",
+    "scifact_json",
+    "scitldr_aic",
+    "scierc_ner",
+    "scierc_relation"
+]
+
+data_dir = Path("/net/nfs.cirrascale/allennlp/davidw/proj/science-instruct/promptsource-sciit/prompts_davidw/tasks")
```
Comment (hamishivi) [inline scripts/prepare_science_data.py]: We should replace this path with a placeholder, and tell folks how to get it.

Comment (hamishivi) [inline scripts/prepare_science_data.py]:
```suggestion
data_dir = Path("data/science/tasks")
```

File: open_instruct/reformat_datasets.py [hunk 4]
```diff
@@ -515,37 +547,243 @@ def convert_open_orca_data(data_dir, output_dir, num_gpt4_examples=100000, num_g
...
+    # we filter out conversations that contain some specific strings
+    filter_strings = [
+        "OpenAI",
+        "Open AI",
+        "ChatGPT",
+        "Chat GPT",
+        "GPT-3",
+        "GPT3",
+        "GPT 3",
+        "GPT-4"
```
Comment (hamishivi) [inline open_instruct/reformat_datasets.py]:
```suggestion
        "GPT-4",
```

**Pull request merged**